crates/oxc_linter/src/rules/react/jsx_key.rs

```diff
@@ -0,0 +1,395 @@
+    #[diagnostic(severity(warning), help("Missing \"key\" prop for element in iterator."))]
+    MissingKeyPropForElementInIterator(#[label] Span),
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct JsxKey;
+
+declare_oxc_lint!(
+    /// ### What it does
+    ///
+    /// Enforce `key` prop for elements in array
+    ///
+    /// ### Example
+    /// ```javascript
+    /// // Bad
+    /// [1, 2, 3].map(x => <App />);
+    /// [1, 2, 3]?.map(x => <BabelEslintApp />)
+    ///
+    /// // Good
+    /// [1, 2, 3].map(x => <App key={x} />);
+    /// [1, 2, 3]?.map(x => <BabelEslintApp key={x} />)
+    /// ```
+    JsxKey,
+    correctness
+);
+
+impl Rule for JsxKey {
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        match node.kind() {
+            AstKind::ArrayExpressionElement(ArrayExpressionElement::Expression(expr)) => match expr
+            {
+                Expression::JSXElement(jsx_elem) => {
+                    check_jsx_element(node, jsx_elem, ctx);
+                }
+                Expression::JSXFragment(jsx_frag) => {
+                    check_jsx_fragment(node, jsx_frag, ctx);
+                }
+                _ => {}
+            },
+            AstKind::ExpressionStatement(expression_stmt) => match &expression_stmt.expression {
+                Expression::JSXElement(jsx_elem) => {
+                    check_jsx_element(node, jsx_elem, ctx);
+                }
+                Expression::JSXFragment(jsx_frag) => {
+                    check_jsx_fragment(node, jsx_frag, ctx);
+                }
+                _ => {}
+            },
+            AstKind::JSXElement(jsx_elem) => {
+                check_jsx_element(node, jsx_elem, ctx);
+            }
+
+            _ => {}
+        }
+    }
+}
+
+enum InsideArrayOrIterator {
+    Array,
+    Iterator,
+}
+
+fn is_in_array_or_iter<'a, 'b>(
+    node: &'b AstNode<'a>,
+    ctx: &'b LintContext<'a>,
+) -> Option<InsideArrayOrIterator> {
+    let mut node = node;
+
+    loop {
+        let Some(parent) = ctx.nodes().parent_node(node.id()) else {
+            return None;
+        };
+
+        match parent.kind() {
+            AstKind::ArrayExpression(_) => return Some(InsideArrayOrIterator::Array),
+            AstKind::CallExpression(v) => {
+                let callee = &v.callee.without_parenthesized();
+
+                if let Expression::MemberExpression(v) = callee {
+                    if let Some(static_property_name) = v.static_property_name() {
+                        if matches!(
+                            static_property_name,
+                            "map"
+                                | "flatMap"
+                                | "from"
+                                | "forEach"
+                                | "filter"
+                                | "some"
+                                | "every"
+                                | "find"
+                                | "findIndex"
```

Comment:
Is this the same as https://github.com/web-infra-dev/oxc/blob/c95f2e093742290343efcd2120b8359d0b58bb05/crates/oxc_linter/src/rules/eslint/array_callback_return/mod.rs#L229-L244 ?

We can move both similar logic to a common place.

Reply:
good point - actually i realizeed we should only really need to check `map`, `flatMap` and `from` - these are the only mapping ones that are used